point_cloud_client/src/lib.rs

```diff
@@ -14,19 +14,88 @@ enum PointClouds {
 pub struct PointCloudClient {
     point_clouds: PointClouds,
     aabb: Aabb3<f64>,
-    pub num_points_per_batch: usize,
-    pub num_threads: usize,
-    pub buffer_size: usize,
+    num_points_per_batch: usize,
+    num_threads: usize,
+    buffer_size: usize,
 }
 
 impl PointCloudClient {
-    pub fn new(locations: &[String], data_provider_factory: DataProviderFactory) -> Result<Self> {
-        if locations.is_empty() {
+    pub fn bounding_box(&self) -> &Aabb3<f64> {
+        &self.aabb
+    }
+
+    pub fn for_each_point_data<F>(&self, point_query: &PointQuery, mut func: F) -> Result<()>
+    where
+        F: FnMut(PointsBatch) -> Result<()>,
+    {
+        match &self.point_clouds {
+            PointClouds::Octrees(octrees) => {
+                let mut parallel_iterator = ParallelIterator::new(
+                    octrees,
+                    point_query,
+                    self.num_points_per_batch,
+                    self.num_threads,
+                    self.buffer_size,
+                );
+                parallel_iterator.try_for_each_batch(&mut func)?;
+            }
+            PointClouds::S2Cells(s2_cells) => {
+                let mut parallel_iterator = ParallelIterator::new(
+                    s2_cells,
+                    point_query,
+                    self.num_points_per_batch,
+                    self.num_threads,
+                    self.buffer_size,
+                );
+                parallel_iterator.try_for_each_batch(&mut func)?;
+            }
+        }
+        Ok(())
```

Comment:
nit: I think you can remove `?;` from each branch and this `Ok())`?

Reply:
Done